geom/type_geometry_collection.go

```diff
@@ -234,3 +240,134 @@ func (c GeometryCollection) Area() float64 {
 	}
 	return sum
 }
+
+func highestDimensionIgnoreEmpties(geom Geometry) int {
+	// We could simply use Dimension() instead of this function
+	// except for the fact empties can have different dimensionalities.
+	// This function thus exists to treat empties as dimensionality 0.
+	if geom.IsEmpty() {
+		return 0
+	}
+	if !geom.IsGeometryCollection() {
+		return geom.Dimension()
+	}
+	c := geom.AsGeometryCollection()
+	highestDim := 0
+	for _, geom := range c.geoms {
+		highestDim = max(highestDim, highestDimensionIgnoreEmpties(geom))
+	}
+	return highestDim
+}
+
+// Centroid of a GeometryCollection is the centroid of its parts' centroids.
+// It returns true iff the centroid is well defined.
+func (c GeometryCollection) Centroid() (Point, bool) {
+	highestDim, numPoints, sumLength, sumArea, sumX, sumY, valid := c.computeCentroid()
+	var xy XY
+	switch highestDim {
+	case 0:
+		if numPoints == 0 {
+			return Point{}, false // Invalid centroid, highestDim is 0 and numPoints is 0
+		}
+		xy = XY{sumX / float64(numPoints), sumY / float64(numPoints)}
+	case 1:
+		if sumLength == 0 {
+			return Point{}, false // Invalid centroid, highestDim is 1 and sumLength is 0
+		}
+		xy = XY{sumX / sumLength, sumY / sumLength}
+	case 2:
+		if sumArea == 0 {
+			return Point{}, false // Invalid centroid, highestDim is 2 and sumArea is 0
+		}
+		xy = XY{sumX / sumArea, sumY / sumArea}
+	default:
+		panic("Invalid dimensionality in centroid calculation.")
+	}
+
+	return NewPointXY(xy), valid
+}
+
+func (c GeometryCollection) computeCentroid() (highestDim, numPoints int, sumLength, sumArea, sumX, sumY float64, valid bool) {
```

Comment:
The number of return args seems pretty high here. What are your thoughts on defining a helper struct to hold everything?

e.g.

```
type centroidCalc struct {
    highestDim int
    numPoints int
    sumXY XY
    sumLength float64
    sumArea float64
    valid bool
}
```

Reply:
Done, although I've kept validity separate.